File: apps/web/src/stores/upgrade-jobs-store.ts
import { create } from 'zustand'

export type UpgradeStage = 'downloading' | 'verifying' | 'pre_flight' | 'installing' | 'restarting'

export type UpgradeStatus = 'running' | 'succeeded' | 'failed' | 'timeout'

export interface UpgradeJob {
  backup_path: string | null
  error: string | null
  finished_at: string | null
  job_id: string
  server_id: string
  stage: UpgradeStage
  started_at: string
  status: UpgradeStatus
  target_version: string
}

interface UpgradeJobsState {
  clearJob: (serverId: string) => void
  getJob: (serverId: string) => UpgradeJob | undefined
  jobs: Map<string, UpgradeJob>
  setJob: (serverId: string, job: UpgradeJob) => void
  setJobs: (jobs: UpgradeJob[]) => void
}

const AUTO_CLEAR_DELAY = 5000

function isFinished(status: UpgradeStatus): boolean {
  return status === 'succeeded' || status === 'failed'
}

export const useUpgradeJobsStore = create<UpgradeJobsState>()((set, get) => ({
  jobs: new Map(),

  setJob: (serverId: string, job: UpgradeJob) => {
    set((state) => {
      const newJobs = new Map(state.jobs)
      newJobs.set(serverId, job)

      if (isFinished(job.status)) {
        setTimeout(() => {
          const currentJob = get().getJob(serverId)
          if (currentJob?.job_id === job.job_id) {
            get().clearJob(serverId)
          }
        }, AUTO_CLEAR_DELAY)
      }

      return { jobs: newJobs }
    })
  },

  clearJob: (serverId: string) => {
    set((state) => {
      const newJobs = new Map(state.jobs)
      newJobs.delete(serverId)
      return { jobs: newJobs }
    })
  },

  setJobs: (jobs: UpgradeJob[]) => {
    const newJobs = new Map<string, UpgradeJob>()
    for (const job of jobs) {
      newJobs.set(job.server_id, job)

      if (isFinished(job.status)) {
        setTimeout(() => {
          const currentJob = get().getJob(job.server_id)
          if (currentJob?.job_id === job.job_id) {
            get().clearJob(job.server_id)
          }
        }, AUTO_CLEAR_DELAY)
      }
    }
    set({ jobs: newJobs })
  },

  getJob: (serverId: string) => {
    return get().jobs.get(serverId)
  }
}))

File: apps/ios/ServerBee/Services/UpgradeJobsStore.swift
import SwiftUI

/// Holds live agent upgrade jobs pushed over the WebSocket, keyed by server id.
/// Mirrors the web zustand `upgrade-jobs-store`: a full sync replaces the whole
/// map, progress frames merge onto an existing job (never create one), and a
/// result frame upserts the terminal state. Finished jobs (succeeded/failed,
/// but NOT timeout) auto-clear after a short delay, guarded by job id so a newer
/// upgrade isn't wiped by an older job's timer.
@MainActor
@Observable
final class UpgradeJobsStore {
    private(set) var jobs: [String: UpgradeJob] = [:]

    private static let autoClearDelay: Duration = .seconds(5)

    /// The live job for a server, if any.
    func job(forServer serverId: String) -> UpgradeJob? { jobs[serverId] }

    /// Replace the whole map from a full sync, then schedule auto-clear for any
    /// already-finished entries.
    func setJobs(_ list: [UpgradeJob]) {
        jobs = Dictionary(uniqueKeysWithValues: list.map { ($0.serverId, $0) })
        for job in list where job.status.isFinished {
            scheduleAutoClear(serverId: job.serverId, jobId: job.jobId)
        }
    }

    /// Merge a progress frame. Progress alone NEVER creates a job — a client that
    /// missed the full sync would otherwise render a partial job — so this no-ops
    /// when there is no existing entry for the server.
    func applyProgress(serverId: String, jobId: String, targetVersion: String, stage: UpgradeStage) {
        guard let existing = jobs[serverId] else { return }
        jobs[serverId] = UpgradeJob(
            serverId: serverId,
            jobId: jobId,
            targetVersion: targetVersion,
            stage: stage,
            status: .running,
            error: existing.error,
            backupPath: existing.backupPath,
            startedAt: existing.startedAt,
            finishedAt: nil
        )
    }

    /// Upsert a result frame. `stage` falls back to the existing stage then
    /// `.downloading` (parity with the web), `startedAt` is preserved, and a
    /// finished status schedules auto-clear.
    func applyResult(
        serverId: String,
        jobId: String,
        targetVersion: String,
        status: UpgradeStatus,
        stage: UpgradeStage?,
        error: String?,
        backupPath: String?
    ) {
        let existing = jobs[serverId]
        let now = WireDate.string(from: Date())
        jobs[serverId] = UpgradeJob(
            serverId: serverId,
            jobId: jobId,
            targetVersion: targetVersion,
            stage: stage ?? existing?.stage ?? .downloading,
            status: status,
            error: error,
            backupPath: backupPath,
            startedAt: existing?.startedAt ?? now,
            finishedAt: now
        )
        if status.isFinished {
            scheduleAutoClear(serverId: serverId, jobId: jobId)
        }
    }

    /// Clear a finished job after the delay, but only if it's still the same job
    /// (a newer upgrade for the same server must survive).
    private func scheduleAutoClear(serverId: String, jobId: String) {
        Task { [weak self] in
            try? await Task.sleep(for: Self.autoClearDelay)
            guard let self, self.jobs[serverId]?.jobId == jobId else { return }
            self.jobs[serverId] = nil
        }
    }
}
